plasmapy/utils/data/downloader.py

```diff
@@ -4,74 +4,409 @@
+                "Unable to connect to data " f"repository {self._API_BASE_URL}"
+            ) from err
 
-        with path.open(mode="wb") as f:
+        # Extract the 'message' value if it is there
+        # If the file does not exist on the repository, the GitHub API
+        # will return `Not Found` in response to this but not raise a 404 error
+        if reply.status_code == 404:
+            raise ValueError(f"URL returned 404: {url}")
+
+        return reply
+
+    def _filepath(self, filename: str) -> Path:
+        """Formats a filepath from a filename."""
+        return Path(self._download_directory, filename)
+
+    def _download_file(self, filename: str, dl_url: str) -> Path:
+        """
+        Download a file from a given URL to a specified path.
+
+        Parameters
+        ----------
+        filename : str
+            Name of the file to download: determines the filepath
+
+        dl_url : str
+            URL from which to download
+
+
+        Returns
+        -------
+        filepath : str
+            Path to the downloaded file
+
+        """
+
+        # Request the contents of the file from the download URL
+        reply = self._http_request(dl_url)
+
+        filepath = self._filepath(filename)
+
+        # Write the contents to file
+        with filepath.open(mode="wb") as f:
             f.write(reply.content)
 
-    return path
+        return filepath
+
+    def _get_file_without_validation(self, filename: str) -> Path:
+        """
+        Return file logic without validation.
+
+        Returns
+        -------
+        filepath : Path
+            Path to file
+
+        Raises
+        ------
+        ValueError
+            If the resource cannot be found locally or on the repository
+
+        """
+        filepath = self._filepath(filename)
+
+        # If the file exists locally, return that
+        if filepath.is_file():
+            return filepath
+
+        # Try blindly downloading from the base URL
+        # Note that downloading directly from the RAW url does not
+        # require an API call.
+        try:
+            dl_url = urljoin(self._RAW_BASE_URL, filename)
+            return self._download_file(filename, dl_url)
+        except ValueError:
+            pass
+
+        raise ValueError(
+            "Resource could not be found locally or "
+            "retrieved from the PlasmPy-data repository: "
+            f"{filename}."
+        )
+
+    def _get_file_with_validation(self, filename: str) -> Path:
+        """
+        Return file logic with validation.
+        """
+        filepath = self._filepath(filename)
+
+        # Try to update the repo blob dict
+        self._update_repo_blob_dict()
+
+        try:
+            local_sha = self._blob_dict[filename]["local_sha"]
+        except KeyError:
+            local_sha = None
+
+        try:
+            repo_sha = self._blob_dict[filename]["repo_sha"]
+        except KeyError:
+            repo_sha = None
```

Comment:
```suggestion
        local_sha = self._blob_dict[filename].get("local_sha", default=None)
        repo_sha = self._blob_dict[filename].get("repo_sha", default=None)
```
This uses `dict.get()` with a default value of `None` if the key isn't present.  I think I got the syntax here right!

Reply:
Oh nice!

Reply:
Syntax is actually dict().(key, default), and defaults to a default of None. 

However, I had to roll this back because the try/except here is actually catching two errors: 

1) If `filename` is not in `_blob_dict`
2) If `local_sha` is not in the `filename` dict. 

And using `get` only handles the first one. Using get twice doesn't work because then it tries to call get a second time on None.